### ISSUE TITLE
Support for out-of-tree vSphere Cloud Provider Interface (CPI) and Cloud Storage Interface (CSI)

### ISSUE THREAD

**Issue by dnoland1** [rancher/rancher#23357]
**What kind of request is this (question/bug/enhancement/feature request):**
Feature request

**Description:**
There is a new out-of-tree vSphere Cloud Provider Interface (CPI), see https://cloud-provider-vsphere.sigs.k8s.io/ . In the k8s 1.20 timeframe, the in-tree cloud providers will be removed and everyone will need to use out-of-tree cloud providers.

The Rancher UI could be enhanced to allow users to select "vSphere" as a Cloud Provider option when creating a custom cluster. User could enter configuration information for vSphere, such as virtual center IP, port, username, password, network, etc. When deploying the cluster, Rancher would automatically create the ConfigMap, Service, DaemonSet, etc. needed for the vsphere-cloud-controller-manager workload.

Support for the vSphere Cloud Storage Interface (CSI) would be great to include as part of this feature request.

This issue may also be relevant: https://github.com/rancher/rancher/issues/20131

gz#6513
gz#9676
gz#12549
gz#12592
gz#14500

**Comment by terafirmanz** [comment]
As part of the move to the cloud controller manager could rancher be extended to run cloud controller manager as part of its own deployment in tree. All k8's clusters would then talk to rancher, rancher would then proxy the requests to the cloud provider.

This would help with clusters that don't have direct connectivity to the cloud provider. It would also make cluster deployment simpler as rancher can auto configure the cluster to point to itself and use stored cloud credential to proxy any requests.

**Comment by Tejeev** [comment]
Our customer would really find the following features useful and it appears they are not available in the in tree providers:
* vm independent volumes
* provisioning form multiple datastores

I couldn't pin down a date or version for the CPI but it looks like both will phase out around k8s v1.21?  This may not be tomorrow but we'll blink and it'll be around the corner so it would be good to have support for this for some time before then so we can shake it down in the field before it's required.

**Comment by Tejeev** [comment]
@cloudnautique Looks like the customer we spoke of is on Openstack and needs CPI and CSI as well.

**Comment by dnoland1** [comment]
This could be useful for those looking for a community contributed Helm chart to deploy vSphere CSI/CPI - https://github.com/stefanvangastel/vsphere-cpi-csi-helm

**Comment by davidnuzik** [comment]
Set milestone, assigned, etc as per Denise.

**Comment by Tejeev** [comment]
I just wanted to make sure it wasn't lost that I was told to use this issue to note the other providers that are needed, rather than new issues.  It is for that reason that I noted OpenStack CPI and CSI here.  @cloudnautique please let me know if this changes and we need to open more issues.

**Comment by Tejeev** [comment]
@deniseschannon will we be able to easily install CSI drivers on a cluster deployed prior to the support thereof?

**Comment by mrajashree** [comment]
Migration can work using the steps from the doc linked above and the chart we're adding in rancher catalog. But due to an existing bug in vsphere CSI driver, it will only work for volumes provisioned using a certain cloud-config format, this issue explains the bug in detail: https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/628

Rancher issue to track it: https://github.com/rancher/rancher/issues/31105

**Comment by sadiapoddar** [comment]
Hi Rajashree,
You mentioned that we can follow the steps from doc link above, I am not see any doc link in your email.
Will you please share the dock link that you are referring to?


Thanks,
-Sadia
________________________________

**Comment by deniseschannon** [comment]
The PR was merged into the branch and is available from the Helm3 library but only for v2.5.6-rc1+ installs. 

We should be including an app-readme to make the experience more user friendly.

**Comment by mrajashree** [comment]
@deniseschannon both charts have an app-readme and a questions.yaml

**Comment by sadiapoddar** [comment]
This bug is blocking me to start working on this issue, I can't create a workload attaching a pv volume.

https://github.com/rancher/rancher/issues/31172

**Comment by nikParasyr** [comment]
The additions do not cover Openstack cpi and csi. Will that be a follow-up or should a separate ticket be created?
Mentioning this as the title includes openstack while description etc does not

**Comment by deniseschannon** [comment]
@nikParasyr This will only cover vSphere and the ticket has been updated to only include that

**Comment by Tejeev** [comment]
@deniseschannon I was told to the CSI/CPI RFE for all drivers into one ticket.  If this has changed we should open another ticket but the priority and history should not be lost as the request has been open for some time.

**Comment by deniseschannon** [comment]
Please note that Rancher does not prohibit out of tree cloud providers at this time, but extending support for Rancher to provide helm charts/ability to easily deploy these cloud providers will need to be per provider. 

Rancher already has the ability to select the "external" cloud provider option and deploy any of these out of tree options.

**Comment by sadiapoddar** [comment]
Verified on 2.5-head, 05fbcbb, k8s 1.19.7 and vSphere version 7.0.x 
**CSI migration from in-tree to out-tree cluster**
 

- Created an in-tree cluster spec ( 1 all role + 1 worker) 
-  Tainted all the nodes by using taints.sh give by Rancher

```
 curl -O https://raw.githubusercontent.com/rancher/helm3-charts/56b622f519728378abeddfe95074f1b87ab73b1e/charts/vsphere-cpi/taints.sh 
  chmod +x taints.sh
  ./taints.sh  .kube/config
```
Once the cluster was up added a workload with a PV volume attached to it
 From helm3-library installed CPI driver app in the cluster 
     - Launch the vsphere-cpi app
     - Installed the app in the existing kube-system namespace
     - Provided the vCenter, data center, and user/pass
     - CPI installation was successful and it created a system pod by the following name
```
vsphere-cloud-controller-manager-xxxx 
```

 Once CPI installation is done made sure all the nodes are marked with the provider information
```
kubectl describe nodes | grep "ProviderID"
```
 From helm3-library installed CSI driver app, vsphere-csi in the cluster 
     - Launch the vsphere-csi app
     - Installed the app in the existing kube-system namespace
     - Provided the vCenter, data center, and user/pass
     - Enable migration to true
     - CSI installation was successful and it created the following pods

```
vsphere-csi-controller-x    
vsphere-csi-node-xxx    
```
 
Enabled feature flags, by editing "Edit as YAML", and  following were addined under kube-controller and kubelet:

```
kube-controller:
  extra_agrs:
      feature-gates: “CSIMigration=true,CSIMigrationvSphere=true”
kubelet:
     extra_agrs:
         feature-gates: “CSIMigration=true,CSIMigration=true"
```

Drained the worker nodes by ending the Cluster YMAL and  selecting "Advanced Options."

- Set the field Maximum Worker Nodes Unavailable to a count of 1.
- To drain the nodes during the upgrade, select Drain Nodes > Yes. 
- Set Force and Delete Local Data to true.
- save the changes.
Once the migration steps were completed, did those to verify the existing PV volume from in-tree can be used

- Created a deployment as daemonset using the existing PV volume 
- Access volume and added some content
- Verified the content consistency by accessing the volume from other worker node workload

Summary: Following the above steps the basic CSI migration worked for me, but need to perform more tests to identify any other issues or corner cases involved with CSI migration. 

Notes: 
1. Used the admin privileges to install CPI and CSI app as I was hitting the following issue and it worked with admin

https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/466#issuecomment-717859491

2. To avoid the following issues, during creating in-tree cluster the cloud-config was given as 

```
default-datastore: <default-datastore-name>
 resourcepool-path: "<cluster-name>/Resources/<resource-pool-name>
```
https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/628

https://github.com/rancher/rancher/issues/31105

**Comment by sadiapoddar** [comment]
Verified 2.5-head, 3e96d7f, k8s 1.20.2 and vSphere 7.0.x 
**Fresh install out-of-tree**

created an out-of-tree cluster by selecting "external" as a cloud provider
cluster spec: 1 node all roles enabled
Verified the cluster YMAL had the following info
```
cloud_provider:
    name: external
```
From helm3-library installed CPI driver app in the cluster
- Launch the vsphere-cpi app
- Installed the app in the existing kube-system namespace
- Provided the vCenter, data center, and user/pass
- CPI installation was successful and it created a system pod by the following name
```
vsphere-cloud-controller-manager-xxxx 
```

 Once CPI installation is done made sure all the nodes are marked with the provider information
```
kubectl describe nodes | grep "ProviderID"
```
 From helm3-library installed CSI driver app, vsphere-csi in the cluster 
     - Launch the vsphere-csi app
     - Installed the app in the existing kube-system namespace
     - Provided the vCenter, data center, and user/pass
     - Enabled migration to false
     - Selected creating a new storage class to true
     - CSI installation was successful and it created the following pods

```
vsphere-csi-controller-x    
vsphere-csi-node-xxx    
```
After the fresh installation is done
 - Verified that a new storage class was created by the name, csi.vsphere.vmware.com
-  Created a workload and attached a new PV volume by using the storage class `csi.vsphere.vmware.com`
-  Verified the storage type is PV volume, not a vmdk disk type
-  Verify the mounted volume is accessible and created some contents in it. 
 
 Summary: Following the above steps the basic out-of-tree fresh install and volume creating with CSI plugin worked. We need to perform more identified tests to find any other issues in an out-of-tree fresh install.

**Comment by mitchellmaler** [comment]
@deniseschannon 
One thing we have noticed using the "external" cloud provider is that it taints all nodes with the "node.cloudprovider.kubernetes.io/uninitialized=true:NoSchedule". Trying to deploy the CSI app/chart using Continuous Delivery fails to start since the fleet agent is stuck Pending after the cluster is brought up. For net new clusters that want to add the CSI App the fleet agent will need tolerations along with launched jobs else this app cannot be installed.

**Comment by bmdepesa** [comment]
@mitchellmaler can you provide some details on what your workflow looks like from the perspective of creating the cluster and deploying the external cloud provider charts? Are you using fleet to deploy the cpi/csi apps?

**Comment by mitchellmaler** [comment]
We use the rancher terraform provider to bring up the vsphere rancher provisioned cluster with the rke config set cloud provider to external which starts the clean cluster with all the nodes tainted. After that we deploy the vsphere cpi/csi first (using manifests or helm cli) so it would register and untaint the nodes. After that all other addons are done using apps v2 (fleet agent helm operation) since it can be scheduled. It would be great to be able to just deploy csi/cpi using the apps v2 as well but since the nodes are all tainted before they can be deployed it causes some workflow issues since fleet does not tolerate those. I haven't thought about using the old apps to deploy this which might solve the problem in our workflow as a temporary workaound if it can run tolerating the taints (will have to give this a try).


A few others potential issues from the pr https://github.com/rancher/helm3-charts/pull/53
1. the resizer container is added by default to csi deployment without a way to disable it. This is only supported on vsphere 7+ and there needs to be a way to toggle it off. (https://github.com/cormachogan/vsphere-csi-helmchart/blob/master/charts/vsphere-csi/values.yaml#L78)
2. I am not sure if the csi node volumes are going to cause an issue since the point to host path "/var/lib/kubelet" where RKE nodes use the path "/opt/rke/var/lib/kubelet"

**Comment by mitchellmaler** [comment]
I guess for #2 that depends on the OS you use such as CoreOS variant or RancherOS. Really the prefix path should be able to be provided in the values to be added to the host paths.

**Comment by bmdepesa** [comment]
rancher/rancher:v2.5.6-rc6
rancher/rancher:master-head 465bb3b

Tested fresh installs, migration, and volume expansion on a vSphere 7.0 environment: 

Fresh install:
* Set cluster's cloud provider to `External`
* Nodes are tainted with `uninitialized`
* Install vSphere CPI chart
* `kubectl describe nodes | grep "ProviderID"` shows the provider IDs correctly and the above taint is removed
* Install vSphere CSI chart
* Use the created storage class to create a PV/PVC and attach to a workload
* Volume is provisioned successfully

Volume Expansion:
* Follow the above steps first
* Remove the workload using the PV (leaving the PV in-place)
* Edit the PV to have a larger size
* See the size reflected in the vSphere volume (Cloud Native Storage) 
* Attach the PV to a workload
* Data persists and the new size is reflected

Migration:
* Create a cluster using the in-tree cloud provider
* Provision a volume using a storage class, attach to a workload
* Create some data in the volume
* Follow the steps from the docs PR: https://github.com/rancher/docs/pull/3000
  * Taint the nodes with the `uninitialized` taint
  * Install the CPI chart, check for `ProviderID` as in fresh install case, and install CSI chart
  * Edit the cluster to add the feature-gates to `kubelet` and `kube-controller` and set drain to true with force and delete local data; Save the updates
  * Cluster is updated and nodes are drained
* When the cluster becomes active, check for migrations with: `kubectl get cnsvspherevolumemigrations`
* Check that the PV and PVC have the `pv.kubernetes.io/migrated-to: csi.vsphere.vmware.com` annotation
* Check the volume attached to the workload and see the data persist

Our QA vSphere 6.7 environment is not at 6.7U3 and as such does not support the out of tree cloud provider, so I was not able to validate the out of tree cloud provider on that environment, however I was able to see the issue mentioned above:
```
the resizer container is added by default to csi deployment without a way to disable it. This is only supported on vsphere 7+ and there needs to be a way to toggle it off. (https://github.com/cormachogan/vsphere-csi-helmchart/blob/master/charts/vsphere-csi/values.yaml#L78)
```

I have logged an issue for this here: https://github.com/rancher/rancher/issues/31550

**Comment by mitchellmaler** [comment]
@bmdepesa how will it deal with the host paths when RKE created nodes that don’t use the standard /var/lib/kubelet paths and have a prefix set?

It is hard coded in the templates paths to use that host path https://github.com/rancher/helm3-charts/blob/429ac83cdb31a87be2d434ac463148cbe9988bc2/charts/vsphere-csi/v2.1.0/templates/vsphere-csi-node-ds.yaml#L122

Flatcar, CoreOS, RancherOS, etc. all use the /opt/rke/var/lib/kubelet paths instead of the standard paths. https://github.com/rancher/rke/blob/master/hosts/hosts.go#L60

The chart should have a way to provide a prefix to all the host paths for those OS types or just for users who set a prefix in their RKE config.

**Comment by bmdepesa** [comment]
Thanks @mitchellmaler 

I was able to confirm this behavior on RancherOS where the `csi-node` workload is failing to deploy due to invalid mount paths. We will be making a change to the chart to support adding the prefix.

**Comment by bmdepesa** [comment]
rancher/rancher:v2.5.6-rc7
rancher-vsphere-cpi:1.0.000-rc01
rancher-vsphere-csi:2.1.000-rc01 

We've moved the charts to the cluster explorer feature charts so they will be bundled in airgap installs, and mirrored all images.

As part of the chart refactoring we exposed the prefix path to `/var/lib/kubelet`. 

* Deployed a RancherOS cluster in vSphere with cloud provider set to `External`
* Deployed vsphere-cpi chart
* Deployed vsphere-csi chart with prefix path set to `/opt/rke`
* Both charts deploy successfully
* Attached a vsphere-csi provisioned volume 

https://github.com/rancher/docs/pull/3090
